### ISSUE TITLE
Curious behavior in the new absent operator with Ruby's `match?`

### ISSUE THREAD

**Issue by hachi8833** [k-takata/Onigmo#86]
First I'd appreciate the release of Ruby 2.4.1 with the new 'absent operator' that comes from Onigmo. Thank you very much! 🎉🎉🎉 

I'd just like to share an amazing/funny behavior of the absent operator `(?~)` on Ruby's `#match?` method. This is just for my interest and I believe it is NOT a bug for now.

*Note: using absent operator for "checking existence of string" to get the result `true`/`false` like the following is **VERY BAD** for practical use. 

```ruby
# always return true
"".match?(/(?~somestring)/)               #=> true
"This is a pen".match?(/(?~somestring)/)  #=> true

# always return false if absent operator contains no regex
"".match?(/(?~)/)                         #=> false
"This is a pen".match?(/(?~)/)            #=> false
```

At first I thought this might be a critical bug, but after that I found an interesting point:

* According to some elemental set theories, "**empty set is a set**" (Perhaps you learned as that in your elementary school age)
* Theoretically, empty strings like **""** can be found among any kind of strings.
  * cf. You can find several empty strings in a string like "abcde": between "a" and "b", "b" and "c" ...etc
* If so, the empty string can be found even in an empty string.
  * in other words, "" contains "", and "" can be found in "".
* As a result, `match?(/(?~somestring)/)` always returns `true` and `match?(/(?~)/)` always returns `false` against any kind of strings including empty string.

If the inference is correct, the behavior that is shown above is consistent in logic. It let me believe that this is not a bug. 
And it suggests again that absent operator should never be used for determining the existence of a string. I think I should be careful when I use absent operators. 

I'd be glad if you find any problems or misunderstanding on that. 

Thank you,

**Comment by nurse** [comment]
absent operator is the one always satisfy following condition.
```ruby
str.match?(reg) != str.match?(/\A(?~#{reg})\z/)
```
This is come from `String#match?` is slightly different one from theoretical regular expression match; it allows partial match. You can create such situation with `\A` and `\z`.

**Comment by k-takata** [comment]
They are expected behavior.
`(?~somestring)` matches "", "s", "so", "som", ... , and "somestrin".
`(?~)` never matches because any string has "" as its substring.
Using absent operator alone might not so useful. You may want to combine with anchors or other strings (prefix and/or suffix).

**Comment by hachi8833** [comment]
Thank you very much for the checks!

FYI: I got the question while I wrote the following article regarding absent operator: 
https://techracho.bpsinc.jp/hachi8833/2017_03_23/37689

**Comment by k-takata** [comment]
> https://techracho.bpsinc.jp/hachi8833/2017_03_23/37689

Thank you for good post.
For Japanese speakers, I also wrote this post: [鬼雲に非包含オペレータを実装した話 - Qiita](http://qiita.com/k-takata/items/4e45121081c83d3d5bfd "鬼雲に非包含オペレータを実装した話 - Qiita").
